java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+          break;
+        default:
+          i++;
+          break;
+      }
+    }
+  }
+
+  private static int parseDelimitersAndContents(
+    String value,
+    int startIndex,
+    int rangeStart,
+    ObjIntConsumer<String> parseContents) {
+    if (startIndex == value.length()) {
+      return startIndex;
+    }
+    var endIndex = parseDelimiterBraces(value, startIndex, rangeStart);
+    if (endIndex == startIndex) {
+      return endIndex;
+    }
+    var contents = value.substring(startIndex + 1, endIndex - 1);
+    parseContents.accept(contents, rangeStart);
+    return endIndex;
+  }
+
+  private static int parseDelimiterBraces(String value, int startIndex, int rangeStart) throws SyntaxError {
+    if (value.charAt(startIndex) != '{') {
+      return startIndex;
+    }
+    var i = startIndex + 1;
+    var openCount = 1;
+
+    while (i < value.length()) {
+      var c = value.charAt(i);
+      i++;
+      switch (c) {
+        case '{':
+          openCount++;
+          break;
+        case '}':
+          openCount--;
+          if (openCount == 0) {
+            return i;
+          }
+          break;
+        default:
+          break;
+      }
+    }
+
+    // +1 because of prefix `$` or `#`
+    var rangeEnd = rangeStart + i - startIndex + 1;
+    throw new SyntaxError("Add missing '}' for this property placeholder or SpEL expression.", rangeStart, rangeEnd);
+  }
+
+  private static void checkValidPropertyPlaceholder(String placeholder, int rangeStart) {
+    if (!isValidPropertyPlaceholder(placeholder, rangeStart)) {
+      // +3 because of delimiter `#{` and `}`
+      var rangeEnd = rangeStart + placeholder.length() + 3;
+      throw new SyntaxError("Correct this malformed property placeholder.", rangeStart, rangeEnd);
+    }
+  }
+
+  private static boolean isValidPropertyPlaceholder(String placeholder, int rangeStart) {
+    var startIndex = 0;
+    var endIndex = placeholder.indexOf(':');
+
+    while (endIndex != -1) {
+      var segment = placeholder.substring(startIndex, endIndex);
+      if (!isValidPropertyPlaceholderSegment(segment, rangeStart + startIndex)) {
+        return false;
+      }
+      startIndex = endIndex + 1;
+      endIndex = placeholder.indexOf(':', startIndex);
+    }
+    var segment = placeholder.substring(startIndex);
+    return isValidPropertyPlaceholderSegment(segment, rangeStart + startIndex);
+  }
+
+  private static boolean isValidPropertyPlaceholderSegment(String segment, int rangeStart) {
+    var stripped = segment.stripLeading();
+    rangeStart += segment.length() - stripped.length();
+    stripped = stripped.stripTrailing();
+
+    if (stripped.startsWith("#{")) {
+      parseDelimitersAndContents(stripped, 1, rangeStart + 2, SpelExpressionCheck::checkValidSpelExpression);
+      return true;
+    } else {
+      return PROPERTY_PLACEHOLDER_PATTERN.matcher(stripped).matches();
```

Comment:
Can't we just use isValidSpelExpression on placeholders?

Reply:
Sadly, no :-/